File: backend-ruby/modules/payments/internal/payment_service.rb
# Business logic — hidden from the outside world.
module Payments
  module Internal
    class PaymentService
      @@store = {}

      def authorize(payment_id, order_id, amount_cents)
        p = { payment_id: payment_id, order_id: order_id, status: "AUTHORIZED", amount_cents: amount_cents }
        @@store[payment_id] = p
        p
      end

      def capture(payment_id)
        p = @@store[payment_id]
        return nil if p.nil? || p[:status] != "AUTHORIZED"
        p[:status] = "CAPTURED"
        p
      end

      def refund(payment_id)
        p = @@store[payment_id]
        return nil if p.nil? || p[:status] != "CAPTURED"
        p[:status] = "REFUNDED"
        p
      end
    end
  end
end

File: backend-kotlin/src/main/kotlin/com/example/modules/payments/internal/PaymentService.kt
// Business logic — hidden from the outside world.
package com.example.modules.payments.internal

import java.util.concurrent.ConcurrentHashMap

data class Payment(val paymentId: String, val orderId: String, var status: String, val amountCents: Long)

class PaymentService {
    private val store = ConcurrentHashMap<String, Payment>()

    fun authorize(paymentId: String, orderId: String, amountCents: Long): Payment {
        val p = Payment(paymentId, orderId, "AUTHORIZED", amountCents)
        store[paymentId] = p
        return p
    }

    fun capture(paymentId: String): Payment? {
        val p = store[paymentId] ?: return null
        if (p.status != "AUTHORIZED") return null
        p.status = "CAPTURED"
        return p
    }

    fun refund(paymentId: String): Payment? {
        val p = store[paymentId] ?: return null
        if (p.status != "CAPTURED") return null
        p.status = "REFUNDED"
        return p
    }
}
